{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# General Python\n",
    "import sys\n",
    "import os\n",
    "sys.path.append('./')\n",
    "\n",
    "\n",
    "# Data structures\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import matplotlib as mpl\n",
    "dpi = 200\n",
    "mpl.rc(\"savefig\", dpi=dpi)\n",
    "%matplotlib inline\n",
    "from scipy.stats.stats import pearsonr\n",
    "from scipy.stats.stats import spearmanr\n",
    "from scipy import stats\n",
    "from scipy.optimize import curve_fit\n",
    "from scipy.optimize import leastsq\n",
    "import seaborn as sns\n",
    "import math\n",
    "import re\n",
    "\n",
    "from lmfit import minimize, Parameters, report_fit\n",
    "from matplotlib.colors import LogNorm\n",
    "\n",
    "#additional imports\n",
    "import pandas as pd\n",
    "from Bio import SeqIO, SeqRecord,Seq\n",
    "import mkl_random\n",
    "import random\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "from Bio import Entrez, SeqIO\n",
    "from copy import deepcopy\n",
    "from pathlib import Path\n",
    "\n",
    "import datetime\n",
    "import time\n",
    "import json\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "from sklearn.metrics import pairwise_distances\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "import seaborn as sb\n",
    "\n",
    "import re\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define a function to return amino acid if given a codon sequence (single amino acid)\n",
    "def get_amino_acid(codon_seq):\n",
    "    from Bio.Seq import Seq\n",
    "    messenger_rna = Seq(codon_seq)\n",
    "    translation=list(messenger_rna.translate())\n",
    "    amino_acid=translation[0]\n",
    "    return (amino_acid)\n",
    "\n",
    "#define a function to translate any DNA sequence\n",
    "def get_translation(codon_seq):\n",
    "    from Bio.Seq import Seq\n",
    "    messenger_rna = Seq(codon_seq)\n",
    "    translation=list(messenger_rna.translate())\n",
    "    return (translation)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define plotting functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "#input desired subsetted mutagenesis window, plate number as ID, and subsetted dataframe\n",
    "#return a histogram representing counts of mutants at each position within the window\n",
    "#and among categories of variants outside the window/not single AA changes\n",
    "def makeVariantHist(windowNum, plateNum, inputDF):\n",
    "    #assign mutational region based on input to imports above (where windowNum is specified)\n",
    "    residue_start = mutationD[windowNum][0]\n",
    "    residue_end=mutationD[windowNum][1]\n",
    "\n",
    "    #make list used for bar chart labels, these are categories that will be grouped from \n",
    "    #multiple residue positions\n",
    "    variant_list=['WT', 'indel','5-UTR','outside']\n",
    "\n",
    "    #create list of individual AA positions that will be counted for this plot\n",
    "    #list of AA residues in window\n",
    "    aa_list= list(mutationD[windowNum][2])\n",
    "    #list of numerical positions of each residue\n",
    "    aa_num_list=range(residue_start, residue_end+1)\n",
    "    #concatenated item by item to give list of aa+pos for all residues in window\n",
    "    res_list= [i + str(j) for i,j in zip (aa_list, aa_num_list)]\n",
    "\n",
    "    #extend grouped residue categories list with individual AAs\n",
    "    variant_list.extend(res_list)\n",
    "\n",
    "    #create new dataframe and add contents of residue/grouped residue containing list\n",
    "    plotDF=pd.DataFrame()\n",
    "    plotDF['Genotype']=variant_list\n",
    "\n",
    "    #make list of all mutated sites\n",
    "    single_sub_sites=[]\n",
    "\n",
    "    #pick out positions from substitution dataframe\n",
    "    for index, val in inputDF.iterrows():\n",
    "        sub=str(val.amino_acid_sub)\n",
    "        codon_num=val.codon\n",
    "\n",
    "        if codon_num==0:\n",
    "            single_sub_sites.append(\"WT\")\n",
    "        if codon_num== -2:\n",
    "            single_sub_sites.append(\"indel\")\n",
    "        if codon_num== -1:\n",
    "            single_sub_sites.append(\"5-UTR\")\n",
    "\n",
    "        if \"_\" in sub:\n",
    "            sub_list=sub.split('_')\n",
    "            aa_num_corrected=int(sub_list[1])\n",
    "            ref_pos=str(sub_list[0])+str(aa_num_corrected)\n",
    "\n",
    "            if ((aa_num_corrected > residue_end) or (aa_num_corrected < residue_start)) and aa_num_corrected >-1:\n",
    "                single_sub_sites.append(\"outside\")\n",
    "\n",
    "            else:\n",
    "                single_sub_sites.append(ref_pos)\n",
    "\n",
    "    #count occurences of each and add to dataframe containing reference seq\n",
    "    count=[]\n",
    "    for index, val in plotDF.iterrows():\n",
    "        num_elements=single_sub_sites.count(val.Genotype)\n",
    "        count.append(num_elements)\n",
    "\n",
    "    plotDF['count']=count\n",
    "    \n",
    "    #make fig and ax\n",
    "    fig=plt.figure(figsize=(20,10))\n",
    "    ax=plt.gca()\n",
    "    \n",
    "    #make bar plot using genotype and count data\n",
    "    ax.bar(list(plotDF['Genotype']),list(plotDF['count']), color='dodgerblue')\n",
    "    \n",
    "    #make string for title\n",
    "    title=\"Plate \" + str(plateNum) + \": Count of SpAP WT, indel and point mutants (n = \"+str(len(inputDF))+\" variants from \"+str(dataDF['i7_barcode'].nunique())+\" barcodes)\"\n",
    "\n",
    "    plt.ylabel('count',fontsize=24)\n",
    "    plt.yticks(np.arange(0, plotDF['count'].max(), 5), fontsize=16)\n",
    "    plt.xlabel('Residue (SpAP window '+str(windowNum)+\", \" +str(residue_start)+\"-\"+str(residue_end)+')',fontsize=24)\n",
    "    plt.xticks(rotation=45, fontsize=16)\n",
    "    plt.title(title, fontsize=24)\n",
    "    plt.tight_layout()\n",
    "    plt.close()\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "## same function as above except return DF containing plotting data only, no fig\n",
    "#input desired subsetted mutagenesis window, plate number as ID, and subsetted dataframe\n",
    "#return a histogram representing counts of mutants at each position within the window\n",
    "#and among categories of variants outside the window/not single AA changes\n",
    "def makeCountDF(windowNum, plateNum, inputDF):\n",
    "    #assign mutational region based on input to imports above (where windowNum is specified)\n",
    "    residue_start = mutationD[windowNum][0]\n",
    "    residue_end=mutationD[windowNum][1]\n",
    "\n",
    "    #make list used for bar chart labels, these are categories that will be grouped from \n",
    "    #multiple residue positions\n",
    "    variant_list=['WT', 'indel','5-UTR','outside']\n",
    "\n",
    "    #create list of individual AA positions that will be counted for this plot\n",
    "    #list of AA residues in window\n",
    "    aa_list= list(mutationD[windowNum][2])\n",
    "    #list of numerical positions of each residue\n",
    "    aa_num_list=range(residue_start, residue_end+1)\n",
    "    #concatenated item by item to give list of aa+pos for all residues in window\n",
    "    res_list= [i + str(j) for i,j in zip (aa_list, aa_num_list)]\n",
    "\n",
    "    #extend grouped residue categories list with individual AAs\n",
    "    variant_list.extend(res_list)\n",
    "\n",
    "    #create new dataframe and add contents of residue/grouped residue containing list\n",
    "    plotDF=pd.DataFrame()\n",
    "    plotDF['Genotype']=variant_list\n",
    "\n",
    "    #make list of all mutated sites\n",
    "    single_sub_sites=[]\n",
    "\n",
    "    #pick out positions from substitution dataframe\n",
    "    for index, val in inputDF.iterrows():\n",
    "        sub=str(val.amino_acid_sub)\n",
    "        codon_num=val.codon\n",
    "\n",
    "        if codon_num==0:\n",
    "            single_sub_sites.append(\"WT\")\n",
    "        if codon_num== -2:\n",
    "            single_sub_sites.append(\"indel\")\n",
    "        if codon_num== -1:\n",
    "            single_sub_sites.append(\"5-UTR\")\n",
    "\n",
    "        if \"_\" in sub:\n",
    "            sub_list=sub.split('_')\n",
    "            aa_num_corrected=int(sub_list[1])\n",
    "            ref_pos=str(sub_list[0])+str(aa_num_corrected)\n",
    "\n",
    "            if ((aa_num_corrected > residue_end) or (aa_num_corrected < residue_start)) and aa_num_corrected >-1:\n",
    "                single_sub_sites.append(\"outside\")\n",
    "\n",
    "            else:\n",
    "                single_sub_sites.append(ref_pos)\n",
    "\n",
    "    #count occurences of each and add to dataframe containing reference seq\n",
    "    count=[]\n",
    "    for index, val in plotDF.iterrows():\n",
    "        num_elements=single_sub_sites.count(val.Genotype)\n",
    "        count.append(num_elements)\n",
    "\n",
    "    plotDF['count']=count\n",
    "    \n",
    "    return plotDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "#input desired subsetted mutagenesis window, plate number as ID, and subsetted dataframe\n",
    "#return a histogram representing counts of mutants at each position within the window\n",
    "#and among categories of variants outside the window/not single AA changes\n",
    "def makeVariantHist_full(inputDF):\n",
    "    #assign mutational region based on input to imports above (where windowNum is specified)\n",
    "    residue_start = 1\n",
    "    residue_end=542\n",
    "\n",
    "    #make list used for bar chart labels, these are categories that will be grouped from \n",
    "    #multiple residue positions\n",
    "    variant_list=['WT', 'indel','5-UTR','outside']\n",
    "\n",
    "    #create list of individual AA positions that will be counted for this plot\n",
    "    #list of AA residues in window\n",
    "    aa_list= protein\n",
    "    #list of numerical positions of each residue\n",
    "    aa_num_list=range(residue_start, residue_end+1)\n",
    "    #concatenated item by item to give list of aa+pos for all residues in window\n",
    "    res_list= [i + str(j) for i,j in zip (aa_list, aa_num_list)]\n",
    "\n",
    "    #extend grouped residue categories list with individual AAs\n",
    "    variant_list.extend(res_list)\n",
    "\n",
    "    #create new dataframe and add contents of residue/grouped residue containing list\n",
    "    plotDF=pd.DataFrame()\n",
    "    plotDF['Genotype']=variant_list\n",
    "\n",
    "    #make list of all mutated sites\n",
    "    single_sub_sites=[]\n",
    "\n",
    "    #pick out positions from substitution dataframe\n",
    "    for index, val in inputDF.iterrows():\n",
    "        sub=str(val.amino_acid_sub)\n",
    "        codon_num=val.codon\n",
    "\n",
    "        if codon_num==0:\n",
    "            single_sub_sites.append(\"WT\")\n",
    "        if codon_num== -2:\n",
    "            single_sub_sites.append(\"indel\")\n",
    "        if codon_num== -1:\n",
    "            single_sub_sites.append(\"5-UTR\")\n",
    "\n",
    "        if \"_\" in sub:\n",
    "            sub_list=sub.split('_')\n",
    "            aa_num_corrected=int(sub_list[1])\n",
    "            ref_pos=str(sub_list[0])+str(aa_num_corrected)\n",
    "\n",
    "            if ((aa_num_corrected > residue_end) or (aa_num_corrected < residue_start)) and aa_num_corrected >-1:\n",
    "                single_sub_sites.append(\"outside\")\n",
    "\n",
    "            else:\n",
    "                single_sub_sites.append(ref_pos)\n",
    "\n",
    "    #count occurences of each and add to dataframe containing reference seq\n",
    "    count=[]\n",
    "    for index, val in plotDF.iterrows():\n",
    "        num_elements=single_sub_sites.count(val.Genotype)\n",
    "        count.append(num_elements)\n",
    "\n",
    "    plotDF['count']=count\n",
    "    \n",
    "    plotDF = plotDF[plotDF.Genotype != 'WT']\n",
    "    plotDF = plotDF[plotDF.Genotype != 'indel']\n",
    "    plotDF = plotDF[plotDF.Genotype != '5-UTR']\n",
    "    plotDF = plotDF[plotDF.Genotype != 'outside']\n",
    "\n",
    "    #make fig and ax\n",
    "    fig=plt.figure(figsize=(60,10))\n",
    "    ax=plt.gca()\n",
    "    \n",
    "    #make bar plot using genotype and count data\n",
    "    ax.bar(list(plotDF['Genotype']),list(plotDF['count']), color='dodgerblue')\n",
    "    \n",
    "    #make string for title\n",
    "    title=\"All single variant mutants across SpAP windows 1-13\"\n",
    "\n",
    "    plt.ylabel('count',fontsize=24)\n",
    "    plt.yticks(np.arange(0, plotDF['count'].max(), 5), fontsize=24)\n",
    "    plt.xlabel('Residue (SpAP All, residues 2-542)',fontsize=18)\n",
    "    plt.xticks(ticks = ['M1','G542'],rotation=45, fontsize=24)\n",
    "    plt.title(title, fontsize=24)\n",
    "    plt.tight_layout()\n",
    "    plt.close()\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Import genome and amino acid sequence for reference + identify residues within each mutational region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import genome (a fasta file with header line and seq line)\n",
    "genome_file=open('/Users/mja/Appel_MutSequencing/snakemake_files/genomes/spap_with_egfp/spap_with_egfp.fa',\"r\")\n",
    "\n",
    "#process file into a list of lines\n",
    "genome_lines=genome_file.readlines()\n",
    "\n",
    "#pull out sequence line as string\n",
    "genome_seq=str(genome_lines[1])\n",
    "\n",
    "#conver string to list for reference\n",
    "genome_seq_list=list(genome_seq)\n",
    "\n",
    "\n",
    "#generate a 'genome' listed in codon format\n",
    "#use this loop to identify absolute base position of first base in start codon for this genome\n",
    "for idx, val in enumerate(genome_seq_list):\n",
    "    start_str=\"ATG\"\n",
    "    \n",
    "    #create string of base and the following two bases\n",
    "    codon_str= str(genome_seq_list[idx])+str(genome_seq_list[idx+1])+str(genome_seq_list[idx+2])\n",
    "    \n",
    "    #check if codon above matches start codon, if so break loop, this is first ATG\n",
    "    if start_str == codon_str:\n",
    "        \n",
    "        #store position of first base in start codon\n",
    "        start_pos = idx\n",
    "        break\n",
    "    \n",
    "#select subset of genome file residing in coding region (here, reference genome\n",
    "#spans 5'UTR, all of SpAP ORF, and some bases beyond SpAP ORF)    \n",
    "genome_seq_coding_list=genome_seq_list.copy()\n",
    "\n",
    "#copy from position of first base in start codon to end\n",
    "genome_seq_coding_list=genome_seq_coding_list[start_pos:len(genome_seq_coding_list)-1]\n",
    "\n",
    "#convert coding list back to a string for splitting into triplets\n",
    "genome_seq_coding_str=''.join(genome_seq_coding_list)\n",
    "\n",
    "#use regex to split string into list of every 3 characters\n",
    "genome_seq_codon_list=re.findall('...', genome_seq_coding_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a reference dictionary containing AA sequence of each mutational region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: [2, 41, 'QSPAPAAAPAPAARSIAATPPKLIVAISVDQFSADLFSEY'],\n",
       " 2: [42, 89, 'RQYYTGGLKRLTSEGAVFPRGYQSHAATETCPGHSTILTGSRPSRTGI'],\n",
       " 3: [90, 137, 'IANNWFDLDAKREDKNLYCAEDESQPGSSSDKYEASPLHLKVPTLGGR'],\n",
       " 4: [138, 185, 'MKAANPATRVVSVAGKDRAAIMMGGATADQVWWLGGPQGYVSYKGVAP'],\n",
       " 5: [186, 232, 'TPLVTQVNQAFAQRLAQPNPGFELPAQCVSKDFPVQAGNRTVGTGRF'],\n",
       " 6: [233, 279, 'ARDAGDYKGFRISPEQDAMTLAFAAAAIENMQLGKQAQTDIISIGLS'],\n",
       " 7: [280, 326, 'ATDYVGHTFGTEGTESCIQVDRLDTELGAFFDKLDKDGIDYVVVLTA'],\n",
       " 8: [311, 356, 'DKLDKDGIDYVVVLTADHGGHDLPERHRMNAMPMEQRVDMALTPKA'],\n",
       " 9: [357, 402, 'LNATIAEKAGLPGKKVIWSDGPSGDIYYDKGLTAAQRARVETEALK'],\n",
       " 10: [403, 448, 'YLRAHPQVQTVFTKAEIAATPSPSGPPESWSLIQEARASFYPSRSG'],\n",
       " 11: [449, 491, 'DLLLLLKPRVMSIPEQAVMGSVATHGSPWDTDRRVPILFWRKG'],\n",
       " 12: [492, 517, 'MQHFEQPLGVETVDILPSLAALIKLP'],\n",
       " 13: [518, 542, 'VPKDQIDGRCLDLVAGKDDSCAGQG'],\n",
       " 14: [518, 542, 'VPKDQIDGRCLDLVAGKDDSCAGQG']}"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a dictionary containing residue boundaries for each mutational region\n",
    "#boundaries are defined with reference to ATG = 1\n",
    "\n",
    "#translate genome (beginning from start codon, as defined above)\n",
    "protein=get_translation(genome_seq_coding_str)\n",
    "\n",
    "#create dict to add boundaries according to region number\n",
    "#boundaries are displayed as a tuple with (leftmost residue #, rightmost residue #)\n",
    "#all subsequent plate numbers match mutational regions for this experiment, except 14\n",
    "#to simplify specify boundaries of 14 (identical to 13)\n",
    "\n",
    "mutationD={1:[2,41],\n",
    "           2:[42,89],\n",
    "           3:[90,137],\n",
    "           4:[138,185],\n",
    "           5:[186,232],\n",
    "           6:[233,279],\n",
    "           7:[280,326],\n",
    "           8:[311,356],\n",
    "           9:[357,402],\n",
    "           10:[403,448],\n",
    "           11:[449,491],\n",
    "           12:[492,517],\n",
    "           13:[518,542],\n",
    "           14:[518,542]\n",
    "          }\n",
    "\n",
    "#iterate through dict to pull out AA seq for each region and store as new item in each entry's list\n",
    "for i in mutationD:\n",
    "    \n",
    "    #get boundaries for region\n",
    "    left_AA_pos=mutationD[i][0]-1 #subtract 1 from left boundary position for 0 indexed list of protein AAs\n",
    "    right_AA_pos=mutationD[i][1] #since non-inclusive, right pos will reference one AA less, don't change\n",
    "    \n",
    "    #pull out region within full list of protein AAs, concatenate to a string\n",
    "    region_list=protein[left_AA_pos:right_AA_pos]   \n",
    "    mut_region_seq = ''.join(region_list) \n",
    "    \n",
    "    \n",
    "    #add this string to dict for each entry\n",
    "    mutationD[i].append(mut_region_seq)\n",
    "mutationD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined=pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_plate_summary_df=pd.DataFrame()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Import csvs from mutant calling notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "#specify window and plate to process\n",
    "expt_windowNumber=1\n",
    "expt_plateNumber=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import initial dataframe created from parsed .vcf file dictionary\n",
    "#this DF contrains columns (refSeq \taltSeq \tfilter \tfw_WT \trv_WT \tfw_alt \trv_alt \tposition \ti7_barcode)\n",
    "#and is dual indexed by 'indel_flag' and 'codon'\n",
    "#it contains all substitutions from all barcodes with variants\n",
    "\n",
    "#specify plate number to analyze\n",
    "# plate_num=1\n",
    "\n",
    "#or, specify list of plate numbers\n",
    "# plate_num_list=range(1,15)\n",
    "\n",
    "#specify location of path containing plate-specific mutant calling CSVs\n",
    "data_path='/Users/mja/Appel_MutSequencing/2021_mut_seq_workup_II/analyses_from_py_notebooks/with_unpaired/mutant_calling/csv_files/plate-well/'\n",
    "plate_path=data_path+\"plate_\"+str(expt_plateNumber)+\"_\"\n",
    "dataDF=pd.read_csv(plate_path+'mutant_calling_initial_dataframe_all_vars.csv')\n",
    "\n",
    "#import indel containing dataframe\n",
    "indels=pd.read_csv(plate_path+'mutant_calling_indels.csv')\n",
    "\n",
    "#import single base subs with amino acid changes added (REMOVE MULTIPLE POSITIONS PER CODON SWAPPED PER BARCODE)\n",
    "df_with_subs_culled_BC_C_index=pd.read_csv(plate_path+'mutant_calling_all_AA_subs.csv')\n",
    "\n",
    "list_of_summary_DFs=[dataDF,indels,df_with_subs_culled_BC_C_index]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate new selected dataframes for plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "#select subset of columns from imported dataframe containing codon sub information (only 1 pos. included per codon\n",
    "#per barcode)\n",
    "single_subs_minimal=df_with_subs_culled_BC_C_index[['i7_barcode',\n",
    "                                                    'codon',\n",
    "                                                    'amino_acid_sub',\n",
    "                                                    'filter',\n",
    "                                                    'fw_WT',\n",
    "                                                    'fw_alt',\n",
    "                                                    'ratio_var',\n",
    "                                                    'plate',\n",
    "                                                    'well'\n",
    "                                                    ]].copy()\n",
    "\n",
    "single_subs_minimal.set_index('i7_barcode', inplace=True)\n",
    "\n",
    "#replace 'none'with -1 and NaN with 0 \n",
    "single_subs_minimal.replace('none',-1, inplace=True)\n",
    "single_subs_minimal.replace('NaN',0, inplace=True)\n",
    "\n",
    "#change data type of codon table to int\n",
    "single_subs_minimal['codon'] = single_subs_minimal['codon'].astype(np.int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_with_subs_culled_BC_C_index=df_with_subs_culled_BC_C_index.loc[df_with_subs_culled_BC_C_index['ratio_var']>100]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Create dataframe showing only barcodes with 1 variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "#now plot only barcodes containing a singlemutant\n",
    "#first, edit dataframe to contain only barcodes that appear once (and with codon substitutions within\n",
    "#range of ORF)\n",
    "single_subs_minimal_reset=single_subs_minimal.reset_index()\n",
    "# subs_indels_empty_reset\n",
    "\n",
    "single_mutantsDF=single_subs_minimal_reset.groupby('i7_barcode').filter(lambda x: len(x) < 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40\n",
      "['I_28_V' 'R_15_V' 'A_8_V' 'K_23_V' 'A_35_V' 'A_9_V' 'P_22_V' 'Q_2_V'\n",
      " 'V_30_A' 'P_12_V' 'Q_32_V' 'A_18_V' 'P_10_V' 'L_24_V' 'P_6_V' 'I_17_V'\n",
      " 'A_11_V' 'F_38_V' 'E_40_V' 'S_16_V' 'P_4_V' 'S_3_V' 'A_5_V' 'A_7_V'\n",
      " 'P_21_V' 'L_37_V' 'A_14_V' 'I_25_V' 'T_20_V' 'S_39_V' 'A_27_V' 'S_29_V'\n",
      " 'Y_41_V' 'D_31_V' 'G_656_S' 'A_13_V' 'F_33_V' 'V_26_A' 'S_39_F' 'R_137_L']\n"
     ]
    }
   ],
   "source": [
    "print(single_mutantsDF['amino_acid_sub'].nunique())\n",
    "print(single_mutantsDF['amino_acid_sub'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create dataframe showing only barcodes with 1 variantnuniqued meeting minimum coverage threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mason/.conda/envs/orthologs/lib/python3.7/site-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "#now plot only barcodes containing a singlemutant\n",
    "#first, edit dataframe to contain only barcodes that appear once (and with codon substitutions within\n",
    "#range of ORF)\n",
    "subs_indels_empty_min_cov_reset=subs_indels_empty_min_cov.reset_index()\n",
    "# subs_indels_empty_reset\n",
    "\n",
    "subs_indels_empty_min_cov_reset_single_var_BCs_only=subs_indels_empty_min_cov_reset.groupby('i7_barcode').filter(lambda x: len(x) < 2)\n",
    "subs_indels_empty_min_cov_reset_single_var_BCs_only['plate_num']=expt_plateNumber"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2628"
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined=pd.concat([combined, subs_indels_empty_min_cov_reset_single_var_BCs_only])\n",
    "\n",
    "len(combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined.to_csv('/media/mason/af107e63-c4f6-4771-8333-2564012a8270/Appel_MutSequencing/mutant_calling/mutant_calling_csvs_FILTER>100/20200121_all_plates_singles_min_cov_only_for_plotting_plate_num.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.1 Plotting histograms per window"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hist 1.1: plot frequency of variants, only show AA subs within desired mutational window \n",
    "#(no coverage theshold)\n",
    "df_1_1=subs_indels_empty.copy()\n",
    "\n",
    "# Hist 1.2: plot frequency of variants, only show AA subs within desired mutational window \n",
    "#(only mutants above read threshold of >95% of positions in genome having at least 1 read\n",
    "df_1_2=subs_indels_empty_min_cov.copy()\n",
    "\n",
    "# Hist 1.3: plot frequency of variants, only show AA subs within desired mutational window \n",
    "#(no coverage theshold) [Barcodes with single mutants only]\n",
    "df_1_3=subs_indels_empty_single_var_BCs_only.copy()\n",
    "\n",
    "# Hist 1.4: plot frequency of variants, only show AA subs within desired mutational window \n",
    "#(only mutants above read threshold of >95% of positions in genome having at least 1 read \n",
    "#[Barcodes with single mutants only]\n",
    "df_1_4=subs_indels_empty_min_cov_reset_single_var_BCs_only.copy()\n",
    "\n",
    "\n",
    "#create dict with names (for saved file) and dataframes to process\n",
    "dictofDFs={\"window_view_df1_all_vars\":df_1_1,\n",
    "           \"window_view_df2_all_vars_mincov\":df_1_2,\n",
    "           \"window_view_df3_singles\":df_1_3,\n",
    "           \"window_view_df4_singles_mincov\":df_1_4\n",
    "          }\n",
    "plotPath=('/media/mason/af107e63-c4f6-4771-8333-2564012a8270/Appel_MutSequencing/mutant_calling/mutant_calling_plots_FILTER>100/bar_plots/')\n",
    "\n",
    "#make bar plot with counts across window for each subsetted dataframe\n",
    "for key, val in dictofDFs.items():\n",
    "    \n",
    "    #create path string for each plate\n",
    "    plotPath_plate=plotPath+\"plate_\"+str(expt_plateNumber)\n",
    "    \n",
    "    #if path does not exist, make it\n",
    "    if os.path.isdir(plotPath_plate) == False:\n",
    "        os.mkdir(plotPath_plate)\n",
    "        \n",
    "    else:\n",
    "        pass\n",
    "    \n",
    "    makeVariantHist(expt_windowNumber,expt_plateNumber,val).savefig(plotPath_plate+\"/plate_\"+str(expt_plateNumber)+\"_window_\"+str(expt_windowNumber)+\"_\"+key+'_filter>100.pdf')\n",
    "    makeVariantHist(expt_windowNumber,expt_plateNumber,val).savefig(plotPath_plate+\"/plate_\"+str(expt_plateNumber)+\"_window_\"+str(expt_windowNumber)+\"_\"+key+'_filter>100.png')\n",
    "    \n",
    "    \n",
    "    #save to joint path for final single var min cov only   \n",
    "    if 'df4' in key:\n",
    "        makeVariantHist(expt_windowNumber,expt_plateNumber,val).savefig(plotPath+\"/plate_\"+str(expt_plateNumber)+\"_window_\"+str(expt_windowNumber)+\"_\"+key+'_filter>100.pdf')\n",
    "        makeVariantHist(expt_windowNumber,expt_plateNumber,val).savefig(plotPath+\"/plate_\"+str(expt_plateNumber)+\"_window_\"+str(expt_windowNumber)+\"_\"+key+'_filter>100.png')\n",
    "    else:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.2 Plotting histograms across entire ORF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [],
   "source": [
    "# makeVariantHist_full(combined).savefig('/media/mason/af107e63-c4f6-4771-8333-2564012a8270/Appel_MutSequencing/mutant_calling/mutant_calling_plots/bar_plots/20200114_all_plates_singles_min_cov_only.pdf')\n",
    "\n",
    "# makeVariantHist_full(combined).savefig('/media/mason/af107e63-c4f6-4771-8333-2564012a8270/Appel_MutSequencing/mutant_calling/mutant_calling_plots/bar_plots/20200114_all_plates_singles_min_cov_only.png')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Organizing summary information for each plate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Pull out counting dataframes from barcodes with any number of variants and barcodes with only one \n",
    "##variant\n",
    "\n",
    "#store df in new variable to reference later, delete rows for non-window positions\n",
    "var_count_mut_window_all=makeCountDF(expt_windowNumber,expt_plateNumber,subs_indels_empty_min_cov)\n",
    "\n",
    "#then drop first 4 rows which list positions not in desired windows\n",
    "var_count_mut_window_all.drop(var_count_mut_window_all.index[[0,1,2,3]], inplace=True)\n",
    "\n",
    "#store df4 in new variable to reference later, delete rows for non-window positions\n",
    "var_count_mut_window_singles=makeCountDF(expt_windowNumber,expt_plateNumber,subs_indels_empty_min_cov_reset_single_var_BCs_only)\n",
    "\n",
    "#then drop first 4 rows which list positions not in desired windows\n",
    "var_count_mut_window_singles.drop(var_count_mut_window_singles.index[[0,1,2,3]], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plate_num</th>\n",
       "      <th>window_num</th>\n",
       "      <th>total_BCs</th>\n",
       "      <th>BCs_processed</th>\n",
       "      <th>BCs_w_min_cov</th>\n",
       "      <th>unique_indels_min_cov</th>\n",
       "      <th>Num_of_BCs_nonIndels_min_cov</th>\n",
       "      <th>num_WT_min_cov</th>\n",
       "      <th>num_singles_min_cov</th>\n",
       "      <th>num_doubles_min_cov</th>\n",
       "      <th>num_triples_plus_min_cov</th>\n",
       "      <th>num_pos_desired_total</th>\n",
       "      <th>num_pos_cov_singles</th>\n",
       "      <th>num_pos_cov_all</th>\n",
       "      <th>ratio_singles_to_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>384</td>\n",
       "      <td>346</td>\n",
       "      <td>340</td>\n",
       "      <td>11</td>\n",
       "      <td>340</td>\n",
       "      <td>35</td>\n",
       "      <td>210</td>\n",
       "      <td>81</td>\n",
       "      <td>14</td>\n",
       "      <td>25</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>0.62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   plate_num  window_num  total_BCs  BCs_processed  BCs_w_min_cov  \\\n",
       "0         14          13        384            346            340   \n",
       "\n",
       "   unique_indels_min_cov  Num_of_BCs_nonIndels_min_cov  num_WT_min_cov  \\\n",
       "0                     11                           340              35   \n",
       "\n",
       "   num_singles_min_cov  num_doubles_min_cov  num_triples_plus_min_cov  \\\n",
       "0                  210                   81                        14   \n",
       "\n",
       "   num_pos_desired_total  num_pos_cov_singles  num_pos_cov_all  \\\n",
       "0                     25                   24               24   \n",
       "\n",
       "   ratio_singles_to_all  \n",
       "0                  0.62  "
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get total number of barcodes processed \n",
    "#find number of unique barcodes in dataframe containg all subs, indels, and empty barcodes\n",
    "index_list=subs_indels_empty.index\n",
    "total_BCs_processed=index_list.nunique()\n",
    "\n",
    "# print(\"number of unique barcodes= \" + str(total_BCs_processed))\n",
    "\n",
    "#get num of barcodes meeting coverage threshold (95% of positions having >=1 read)\n",
    "subs_indels_empty_min_cov_set_index=subs_indels_empty_min_cov.copy()\n",
    "subs_indels_empty_min_cov_set_index.set_index('i7_barcode',inplace=True)\n",
    "index_list_min_cov=subs_indels_empty_min_cov_set_index.index\n",
    "total_BCs_min_cov=index_list_min_cov.nunique()\n",
    "\n",
    "#get number of barcodes containing: indels\n",
    "list_of_unique_indels_min_cov=[]\n",
    "\n",
    "for index,val in indels_minimal.iterrows():\n",
    "    if index in index_list_min_cov:\n",
    "        list_of_unique_indels_min_cov.append(index)\n",
    "    else:\n",
    "        pass\n",
    "num_unique_BC_indels_min_cov=len(list_of_unique_indels_min_cov)\n",
    "\n",
    "####putting indels aside, how many WT, single mutants, doubles, triple or more?\n",
    "##first drop duplicates\n",
    "#use this line to create a new dataframe column in copied df\n",
    "#and add frequency of each codon to it\n",
    "subs_indels_empty_min_cov_count=subs_indels_empty_min_cov.copy()\n",
    "subs_indels_empty_min_cov_count['i7_barcode_freq'] = subs_indels_empty_min_cov_count.groupby('i7_barcode')['i7_barcode'].transform('count')\n",
    "subs_indels_empty_min_cov_count_dropped=subs_indels_empty_min_cov_count.copy()\n",
    "subs_indels_empty_min_cov_count_dropped=subs_indels_empty_min_cov_count_dropped.reset_index().drop_duplicates(subset='i7_barcode', keep='first').set_index('i7_barcode')\n",
    "\n",
    "\n",
    "\n",
    "##num of all non-indels with minimum coverage (note, some of these will have associated indels as well)\n",
    "#codon designation for indel is -2\n",
    "num_non_indels_min_cov=len(subs_indels_empty_min_cov_count_dropped[subs_indels_empty_min_cov_count_dropped.codon != -2])\n",
    "\n",
    "##Now use this df to calculate number of WT, singles, doubles, triples and more\n",
    "#codon designation for WT is 0\n",
    "num_WT_min_cov=len(subs_indels_empty_min_cov_count_dropped[subs_indels_empty_min_cov_count_dropped.codon == 0])\n",
    "\n",
    "#codon designation for \n",
    "num_all_mut_min_cov=len(subs_indels_empty_min_cov_count_dropped[subs_indels_empty_min_cov_count_dropped.codon >= 1])\n",
    "\n",
    "#get single mutants using this selection format\n",
    "num_singles_min_cov=len(subs_indels_empty_min_cov_count_dropped[(subs_indels_empty_min_cov_count_dropped.codon != -2) & (subs_indels_empty_min_cov_count_dropped.codon != 0) & (subs_indels_empty_min_cov_count_dropped.i7_barcode_freq == 1)])\n",
    "\n",
    "#get doubles mutants using this selection format\n",
    "num_doubles_min_cov=len(subs_indels_empty_min_cov_count_dropped[(subs_indels_empty_min_cov_count_dropped.codon != -2) & (subs_indels_empty_min_cov_count_dropped.codon != 0) & (subs_indels_empty_min_cov_count_dropped.i7_barcode_freq == 2)])\n",
    "\n",
    "#get triples mutants using this selection format\n",
    "num_triples_plus_min_cov=len(subs_indels_empty_min_cov_count_dropped[(subs_indels_empty_min_cov_count_dropped.codon != -2) & (subs_indels_empty_min_cov_count_dropped.codon != 0) & (subs_indels_empty_min_cov_count_dropped.i7_barcode_freq >= 3)])\n",
    "\n",
    "####number of desired positions represented in barcodes with just one variant\n",
    "##use DF copied from section 1.4 and use selection format to find positions with counts !=0\n",
    "num_pos_desired_total=len(var_count_mut_window_singles)\n",
    "#rename count column because cannot reference it without getting confused for a counting method\n",
    "# var_count_mut_window_singles.columns = ['number']\n",
    "\n",
    "var_count_mut_window_singles.rename(columns={'count':'number'}, inplace=True)\n",
    "num_pos_cov_singles=len(var_count_mut_window_singles[var_count_mut_window_singles.number > 0])\n",
    "\n",
    "\n",
    "\n",
    "#####number of desired positions represented in all barcodes\n",
    "# var_count_mut_window_all.columns = ['number']\n",
    "\n",
    "var_count_mut_window_all.rename(columns={'count':'number'}, inplace=True)\n",
    "\n",
    "num_pos_cov_all=len(var_count_mut_window_all[var_count_mut_window_all.number > 0])\n",
    "\n",
    "\n",
    "\n",
    "#list of barcodes with single variants, and their codon change\n",
    "\n",
    "#best single variant (one with most reads/highest coverage)\n",
    "\n",
    "#list of positions not covered\n",
    "\n",
    "\n",
    "#list of barcodes from multi mutants containing missing positions\n",
    "\n",
    "plate_summary_df=pd.DataFrame({'plate_num':[expt_plateNumber],\n",
    "                                'window_num':[expt_windowNumber],\n",
    "                                'total_BCs':[384], \n",
    "                                'BCs_processed':[total_BCs_processed],\n",
    "                                'BCs_w_min_cov':[total_BCs_min_cov],\n",
    "                                'unique_indels_min_cov':[num_unique_BC_indels_min_cov],\n",
    "                                'Num_of_BCs_nonIndels_min_cov':[num_non_indels_min_cov],\n",
    "                                'num_WT_min_cov':[num_WT_min_cov],\n",
    "                                'num_singles_min_cov':[num_singles_min_cov],\n",
    "                                'num_doubles_min_cov':[num_doubles_min_cov],\n",
    "                                'num_triples_plus_min_cov':[num_triples_plus_min_cov],\n",
    "                                'num_pos_desired_total':[num_pos_desired_total],\n",
    "                                'num_pos_cov_singles':[num_pos_cov_singles],\n",
    "                                'num_pos_cov_all':[num_pos_cov_all],\n",
    "                               'ratio_singles_to_all':[round((num_singles_min_cov/total_BCs_min_cov),2)]\n",
    "                               \n",
    "                              })\n",
    "\n",
    "#export summary dataframe to csv\n",
    "plate_summary_df.to_csv('/media/mason/af107e63-c4f6-4771-8333-2564012a8270/Appel_MutSequencing/mutant_calling/mutant_calling_summary_stats_csvs_FILTER>100/plate_'+str(expt_plateNumber)+'_mutant_stats.csv')\n",
    "plate_summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_plate_summary_df=pd.concat([combined_plate_summary_df, plate_summary_df])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_plate_summary_df\n",
    "\n",
    "combined_plate_summary_df.to_csv('/media/mason/af107e63-c4f6-4771-8333-2564012a8270/Appel_MutSequencing/mutant_calling/mutant_calling_summary_stats_csvs_FILTER>100/plates_1-14_mutant_stats.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
