{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 394,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !/usr/bin/env python\n",
    "\n",
    "# [Future imports]\n",
    "# \"print\" function compatibility between Python 2.x and 3.x\n",
    "from __future__ import print_function\n",
    "# Use Python 3.x \"/\" for division in Pyhton 2.x\n",
    "from __future__ import division\n",
    "\n",
    "# General Python\n",
    "import sys\n",
    "import os\n",
    "sys.path.append('./')\n",
    "# Data structures\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import matplotlib as mpl\n",
    "dpi = 200\n",
    "mpl.rc(\"savefig\", dpi=dpi)\n",
    "%matplotlib inline\n",
    "from scipy.stats.stats import pearsonr\n",
    "from scipy.stats.stats import spearmanr\n",
    "from scipy import stats\n",
    "from scipy.optimize import curve_fit\n",
    "from scipy.optimize import leastsq\n",
    "import seaborn as sns\n",
    "import math\n",
    "import re\n",
    "\n",
    "from lmfit import minimize, Parameters, report_fit\n",
    "from matplotlib.colors import LogNorm\n",
    "\n",
    "#additional imports\n",
    "import pandas as pd\n",
    "from Bio import SeqIO, SeqRecord,Seq\n",
    "import mkl_random\n",
    "import random\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "from Bio import Entrez, SeqIO\n",
    "from copy import deepcopy\n",
    "from pathlib import Path\n",
    "\n",
    "import datetime\n",
    "import time\n",
    "import json\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "from sklearn.metrics import pairwise_distances\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import re\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Define functions used in this notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define a function to convert a base position within the ORF of a variant \n",
    "#to a codon position\n",
    "\n",
    "def get_codon(pos):\n",
    "\n",
    "    import math\n",
    "\n",
    "    #subtract 1 to match index of genome list\n",
    "    reindex_pos=pos-1\n",
    "    \n",
    "    #pass if start codon is not within ORF\n",
    "    if reindex_pos < start_pos:\n",
    "        return 'none'\n",
    "    \n",
    "    #else, subtract start position from position of interest and round upward\n",
    "    else:\n",
    "        codon = math.ceil((reindex_pos-start_pos+1)/3)\n",
    "        return codon\n",
    "    \n",
    "    \n",
    "#define a function to identify the position of the stop codon within a list of\n",
    "#codons for a protein: the codon list begins with ATG at index 0\n",
    "def get_stop_codon_pos(codon_list):\n",
    "    for index, codon in enumerate(codon_list):\n",
    "        if codon == 'TAA':\n",
    "            stop_codon_pos= index #position of stop codon in codon list where item 0 = start codon ATG\n",
    "            break\n",
    "    return(stop_codon_pos)\n",
    "\n",
    "\n",
    "#define a function to return amino acid if given a codon sequence (single amino acid)\n",
    "def get_amino_acid(codon_seq):\n",
    "    from Bio.Seq import Seq\n",
    "    messenger_rna = Seq(codon_seq)\n",
    "    translation=list(messenger_rna.translate())\n",
    "    amino_acid=translation[0]\n",
    "    return (amino_acid)\n",
    "\n",
    "#define a function to translate any DNA sequence\n",
    "def get_translation(codon_seq):\n",
    "    from Bio.Seq import Seq\n",
    "    messenger_rna = Seq(codon_seq)\n",
    "    translation=list(messenger_rna.translate())\n",
    "    return (translation)\n",
    "\n",
    "#define function to return codon change of single base substitutions\n",
    "def get_var_change(pos_lists, pos_identity, codon_pos):\n",
    "    #codon pos of -1 means mutatation is located in genome at a position outside of ORF (defined in\n",
    "    #get codon function)  \n",
    "    if codon_pos[0]==\"none\":\n",
    "        return 'NaN'\n",
    "    else:\n",
    "        codon_pos_val=codon_pos[0]-1\n",
    "\n",
    "        #get codon start position (with respect to 0 indexed genome list)\n",
    "        codon_start_pos=math.trunc(((codon_pos_val)*3)+start_pos)\n",
    "\n",
    "        #this must be re-indexed for -1, ie, genome_seq_codon_list[0]=ATG\n",
    "        WT_codon=''.join((genome_seq_list[codon_start_pos:(codon_start_pos+3)]))\n",
    "\n",
    "        #get codon start position \n",
    "        #(with respect to 1 indexed variant position information from vcf/dataframe)\n",
    "        codon_start_pos_vcf_index=codon_start_pos+1\n",
    "\n",
    "        #get WT amino acid\n",
    "        WT_amino_acid=get_amino_acid(WT_codon)\n",
    "\n",
    "\n",
    "        #assign position index to each position in WT codon using dictionary\n",
    "        WT_codonDict={}\n",
    "        WT_codonDict[codon_start_pos_vcf_index]=WT_codon[0]\n",
    "        WT_codonDict[codon_start_pos_vcf_index+1]=WT_codon[1]\n",
    "        WT_codonDict[codon_start_pos_vcf_index+2]=WT_codon[2]\n",
    "\n",
    "        #create dictionary for substitutions in codon\n",
    "        #zip function accepts list of keys and list of values to do this\n",
    "        var_codonDict= dict(zip(pos_lists, pos_identity))\n",
    "\n",
    "        #now, create modified dictionary containing variant codon\n",
    "        #which may include 1 or 2 bases from the WT codon\n",
    "\n",
    "        WT_codonDict.update(var_codonDict)\n",
    "\n",
    "        #Extract full variant codon in string format in order to translate to amino acid\n",
    "\n",
    "        #create string to store full variant codon\n",
    "        var_codon=\"\"\n",
    "        for key, value in WT_codonDict.items():\n",
    "            var_codon = var_codon+value\n",
    "\n",
    "        #translate variant codon to amino acid residue\n",
    "        var_amino_acid=get_amino_acid(var_codon)\n",
    "\n",
    "        #create string summarizing amino acid mutation\n",
    "        var_change=WT_amino_acid+\"_\"+str(codon_pos[0])+\"_\"+var_amino_acid\n",
    "\n",
    "\n",
    "        #return this variable to store to 'newseq' in get_amino_acid_change\n",
    "        return var_change\n",
    "\n",
    "#define function to get amino acid change (could probably do this function and above in one instead)\n",
    "#input dataframe for this function is the \"barcode_codon_index\" grouped dataframe\n",
    "#therefore, only substitions within the same codon will be processed together\n",
    "def get_amino_acid_change(df):\n",
    "    \n",
    "    newdf = df.copy()\n",
    "    \n",
    "    #get values from relevant columns\n",
    "    codon_pos=df.codon.values\n",
    "    pos_lists=list(df.position.values)\n",
    "    pos_identity=list(df.altSeq.values)\n",
    "    \n",
    "    #call get_var_change function\n",
    "    newseq = get_var_change(pos_lists, pos_identity, codon_pos)\n",
    "    newdf['amino_acid_sub'] = newseq\n",
    "    #return dataframe with additional column\n",
    "    return newdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Define variant call directories and import corresponding genome fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 396,
   "metadata": {},
   "outputs": [],
   "source": [
    "##directory for variant call files (purexpress entire genome)\n",
    "dirName_1to13_purexp_cov = '/Users/mja/Appel_MutSequencing/2021_mut_seq_workup_II/variant_calls/20201222_SU011_MJA_1-13_with_R1_R2_variant_calls'\n",
    "dirName_14_purexp_cov = '/Users/mja/Appel_MutSequencing/2021_mut_seq_workup_II/variant_calls/20201219_SU012_MJA_14_with_R1_R2_variant_calls'\n",
    "\n",
    "#generate list of files in directory\n",
    "file_1to13_purexp = os.listdir(dirName_1to13_purexp_cov)\n",
    "file_14_purexp = os.listdir(dirName_14_purexp_cov)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import genome (a fasta file with header line and seq line)\n",
    "genome_file=open('/Users/mja/Appel_MutSequencing/snakemake_files/genomes/spap_with_egfp/spap_with_egfp.fa',\"r\")\n",
    "\n",
    "#process file into a list of lines\n",
    "genome_lines=genome_file.readlines()\n",
    "\n",
    "#pull out sequence line as string\n",
    "genome_seq=str(genome_lines[1])\n",
    "\n",
    "#conver string to list for reference\n",
    "genome_seq_list=list(genome_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 398,
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate a 'genome' listed in codon format\n",
    "\n",
    "#use this loop to identify absolute base position of first base in start codon for this genome\n",
    "for idx, val in enumerate(genome_seq_list):\n",
    "    start_str=\"ATG\"\n",
    "    \n",
    "    #create string of base and the following two bases\n",
    "    codon_str= str(genome_seq_list[idx])+str(genome_seq_list[idx+1])+str(genome_seq_list[idx+2])\n",
    "    \n",
    "    #check if codon above matches start codon, if so break loop, this is first ATG\n",
    "    if start_str == codon_str:\n",
    "        \n",
    "        #store position of first base in start codon\n",
    "        start_pos = idx\n",
    "        break\n",
    "    \n",
    "#select subset of genome file residing in coding region (here, reference genome\n",
    "#spans 5'UTR, all of SpAP ORF, and some bases beyond SpAP ORF)    \n",
    "genome_seq_coding_list=genome_seq_list.copy()\n",
    "\n",
    "#copy from position of first base in start codon to end\n",
    "genome_seq_coding_list=genome_seq_coding_list[start_pos:len(genome_seq_coding_list)-1]\n",
    "\n",
    "\n",
    "#convert coding list back to a string for splitting into triplets\n",
    "genome_seq_coding_str=''.join(genome_seq_coding_list)\n",
    "\n",
    "#use regex to split string into list of every 3 characters\n",
    "genome_seq_codon_list=re.findall('...', genome_seq_coding_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Process entire directory of vcf files and ouput dictionary with variant data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 399,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4646"
      ]
     },
     "execution_count": 399,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# let's make a dictionary for data and read in the data from the files\n",
    "dirName=dirName_1to13_purexp_cov\n",
    "fileL=file_1to13_purexp\n",
    "\n",
    "dataD = {}\n",
    "\n",
    "for a in fileL:\n",
    "    #get file name without extension\n",
    "    name=a.split('.')\n",
    "     \n",
    "    #pull out i7+i5\n",
    "    i7=(name[0].split('-'))[3]\n",
    "   \n",
    "    i5=((name[0].split('-'))[7]).split('_')[0]\n",
    "    \n",
    "    # let's create a dictionary entry for that barcode name (i7)\n",
    "    bcName=int(i7)\n",
    "    \n",
    "    dataD[bcName] = {}\n",
    "    \n",
    "    # open each file\n",
    "    f = open(dirName + '/'+a,'r')\n",
    "    \n",
    "    #read it line by line\n",
    "    for line in f:\n",
    "        # ignore it if it starts with a #\n",
    "        if line[0] == '#':\n",
    "            pass\n",
    "        else:\n",
    "            # create a tab-delimited list of things\n",
    "            tempL = line.split('\\t')\n",
    "            # get the position number of a potential variant\n",
    "            posNum = int(tempL[1])\n",
    "            # get the reference sequence at that position\n",
    "            refSeq = tempL[3]\n",
    "            # get the variant sequence at that position\n",
    "            altSeq = tempL[4]\n",
    "            # get the filter score\n",
    "            filterScore = float(tempL[5])\n",
    "            \n",
    "            \n",
    "            #get codon number from reference sequence\n",
    "            codonNum=get_codon(posNum)\n",
    "            \n",
    "            #create tuple to uniquely identify variants even if they share the same position\n",
    "            pos_unique=(posNum, refSeq, altSeq)\n",
    "            \n",
    "            \n",
    "            # find the entry that has the DP4 stuff with FW and RV reference and alternate read counts\n",
    "            # this list has lots of stuff at variable spots so be careful!\n",
    "            varStats = tempL[7].split(';')\n",
    "            for b in varStats:\n",
    "                # identify the thing in the list that looks like DP4 = \n",
    "                if 'DP4' in b:\n",
    "                    # now get the reads\n",
    "                    fw_WT = int(b.split('=')[1].split(',')[0])\n",
    "                    rv_WT = int(b.split('=')[1].split(',')[1])\n",
    "                    fw_alt = int(b.split('=')[1].split(',')[2])\n",
    "                    rv_alt = int(b.split('=')[1].split(',')[3])\n",
    "            # now add to nested dictionary                                                                        \n",
    "            dataD[bcName][pos_unique] = {}\n",
    "            dataD[bcName][pos_unique]['refSeq'] = refSeq\n",
    "            dataD[bcName][pos_unique]['altSeq'] = altSeq\n",
    "            dataD[bcName][pos_unique]['filter'] = filterScore\n",
    "            dataD[bcName][pos_unique]['fw_WT'] = fw_WT\n",
    "            dataD[bcName][pos_unique]['rv_WT'] = rv_WT\n",
    "            dataD[bcName][pos_unique]['fw_alt'] = fw_alt\n",
    "            dataD[bcName][pos_unique]['rv_alt'] = rv_alt\n",
    "            dataD[bcName][pos_unique]['codon']=codonNum\n",
    "            dataD[bcName][pos_unique]['position']=posNum\n",
    "            dataD[bcName][pos_unique]['i7_barcode']=bcName\n",
    "    \n",
    "    #close file\n",
    "    f.close()\n",
    "    \n",
    "len(dataD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Data filtering step: identify all i7 barcodes containing 10X read depth at >= 95% of positions (and at least 1 read at all positions)\n",
    "-Coverage stat dataframes contain this information, import and make a list of barcodes failing to meet this depth threshold  \n",
    "-In next step, converting dictionary to dataframe, remove these barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 400,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3530"
      ]
     },
     "execution_count": 400,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#import csv containing depth stat information (from notebook 'II_Coverage_Stats')\n",
    "coverageDF = pd.read_csv('/Users/mja/Appel_MutSequencing/2021_mut_seq_workup_II/analyses_from_py_notebooks/with_unpaired/depth_results/20210602_plates_1-13_read_stats_allgenomes_cov_stats.csv')\n",
    "coverageDF.drop(columns=['Unnamed: 0'], inplace=True)\n",
    "\n",
    "#convert nan to zeros\n",
    "coverageDF.fillna(0, inplace=True)\n",
    "\n",
    "#identify entries with less than 10x depth at 95+% of positions\n",
    "barcodesToKeepDF=coverageDF.loc[(coverageDF['frac_pos_cov>=1'] == 1.00) & (coverageDF['frac_pos_cov>=10'] > 0.94)]\n",
    "\n",
    "barcodesKeepL=barcodesToKeepDF['i7'].tolist()\n",
    "\n",
    "len(barcodesKeepL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Convert dictionary (for barcodes containing variants only) to dataframe\n",
    "-Remove all barcodes not in keep list from step 4  \n",
    "-after that, find empty vcf files, these are WT  \n",
    "-Results 6/3/21:  \n",
    "-3530 meeting depth threshold, 327 are WT, 3203 have variants (of any type)  \n",
    "-Of 3203, 174 barcodes have indels (at least 1 indel, and any other type of sub), and 3029 have only single base substitutions (but could be single, double, triple or higher mutants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 401,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refSeq</th>\n",
       "      <th>altSeq</th>\n",
       "      <th>filter</th>\n",
       "      <th>fw_WT</th>\n",
       "      <th>rv_WT</th>\n",
       "      <th>fw_alt</th>\n",
       "      <th>rv_alt</th>\n",
       "      <th>position</th>\n",
       "      <th>i7_barcode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>codon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>CT</td>\n",
       "      <td>CTT</td>\n",
       "      <td>43.9865</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>576</td>\n",
       "      <td>3666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>582</th>\n",
       "      <td>CGGCGAGGGCGAGGGCGA</td>\n",
       "      <td>CGGCGAGGGCGA</td>\n",
       "      <td>221.0000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1818</td>\n",
       "      <td>1142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>223.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>872</td>\n",
       "      <td>4324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>490</th>\n",
       "      <td>AGG</td>\n",
       "      <td>AG</td>\n",
       "      <td>225.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>1542</td>\n",
       "      <td>6938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ATT</td>\n",
       "      <td>AT</td>\n",
       "      <td>225.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>7</td>\n",
       "      <td>203</td>\n",
       "      <td>1242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>GTTTT</td>\n",
       "      <td>GTTT</td>\n",
       "      <td>225.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>69</td>\n",
       "      <td>82</td>\n",
       "      <td>1530</td>\n",
       "      <td>4412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>TGA</td>\n",
       "      <td>T</td>\n",
       "      <td>225.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>25</td>\n",
       "      <td>1142</td>\n",
       "      <td>6025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>CG</td>\n",
       "      <td>C</td>\n",
       "      <td>225.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>680</td>\n",
       "      <td>3941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CCCTGCGGCACGTTCCATCGCAGCTACGCCTCCTAAACTGATCGTG...</td>\n",
       "      <td>C</td>\n",
       "      <td>30.4183</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>105</td>\n",
       "      <td>860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547</th>\n",
       "      <td>AGG</td>\n",
       "      <td>AGGG</td>\n",
       "      <td>225.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>1713</td>\n",
       "      <td>308</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>240 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  refSeq        altSeq  \\\n",
       "codon                                                                    \n",
       "168                                                   CT           CTT   \n",
       "582                                   CGGCGAGGGCGAGGGCGA  CGGCGAGGGCGA   \n",
       "267                                                   AG             A   \n",
       "490                                                  AGG            AG   \n",
       "44                                                   ATT            AT   \n",
       "...                                                  ...           ...   \n",
       "486                                                GTTTT          GTTT   \n",
       "357                                                  TGA             T   \n",
       "203                                                   CG             C   \n",
       "11     CCCTGCGGCACGTTCCATCGCAGCTACGCCTCCTAAACTGATCGTG...             C   \n",
       "547                                                  AGG          AGGG   \n",
       "\n",
       "         filter  fw_WT  rv_WT  fw_alt  rv_alt  position  i7_barcode  \n",
       "codon                                                                \n",
       "168     43.9865      6      5       2       4       576        3666  \n",
       "582    221.0000      2      0       4       3      1818        1142  \n",
       "267    223.0000      0      0      14       6       872        4324  \n",
       "490    225.0000      0      0      29      29      1542        6938  \n",
       "44     225.0000      0      0      16       7       203        1242  \n",
       "...         ...    ...    ...     ...     ...       ...         ...  \n",
       "486    225.0000      0      0      69      82      1530        4412  \n",
       "357    225.0000      0      0      42      25      1142        6025  \n",
       "203    225.0000      0      0      15      14       680        3941  \n",
       "11      30.4183      0      0       1       0       105         860  \n",
       "547    225.0000      0      0      13       9      1713         308  \n",
       "\n",
       "[240 rows x 9 columns]"
      ]
     },
     "execution_count": 401,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#convert dictionary into dataframe for downstream steps\n",
    "dataframeflatdict = []\n",
    "BCs_with_empty_vcf=[]\n",
    "skipped=[]\n",
    "\n",
    "#create flat dictionary from nested dictionary above\n",
    "for k, v in dataD.items():\n",
    "\n",
    "    if k in barcodesKeepL:\n",
    "        #identify barcodes with no entries (vcf files empty)\n",
    "        #if v: evaluates to 'True' then pass(there are variants here)\n",
    "        if v:\n",
    "            pass\n",
    "\n",
    "        #if no variants, add barcode to list to reference later\n",
    "        else:\n",
    "            BCs_with_empty_vcf.append(k)\n",
    "\n",
    "        for entrykey, entryvalue in v.items():\n",
    "            dataframeflatdict.append(entryvalue) # entryvalue is a dictionary, the innermost\n",
    "    else:\n",
    "        skipped.append(k)\n",
    "        \n",
    "#make new dataframe from flattened dictionary\n",
    "dataDF=pd.DataFrame(dataframeflatdict)\n",
    "\n",
    "#add \"indel flag\" if variant is not the same length as reference sequence\n",
    "dataDF['indel_flag'] = (dataDF.altSeq.str.len() != 1) | (dataDF.refSeq.str.len() != 1)\n",
    "dataDF.set_index(['indel_flag','codon'], inplace = True)\n",
    "dataDF['i7_barcode'].nunique()\n",
    "\n",
    "dataDF.loc[True]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5.1 Export DFs containing barcodes tossed for failing depth thresholds, and WT barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 402,
   "metadata": {},
   "outputs": [],
   "source": [
    "#add list of skipped barcodes to dataframe and export for later reference\n",
    "#these are barcodes NOT meeting depth thresholds -- cannot be categorized as WT or variant\n",
    "(pd.DataFrame({'i7_barcode':skipped})).to_csv('/Users/mja/Appel_MutSequencing/2021_mut_seq_workup_II/analyses_from_py_notebooks/with_unpaired/mutant_calling/csv_files/20210603_plates_1-13_barcodes_depth_failed_'+str(len(skipped))+'.csv')\n",
    "\n",
    "#add list of WT barcodes to dataframe and export for later reference\n",
    "#these are barcodes meeting depth thresholds but having empty vcf files\n",
    "(pd.DataFrame({'i7_barcode':BCs_with_empty_vcf})).to_csv('/Users/mja/Appel_MutSequencing/2021_mut_seq_workup_II/analyses_from_py_notebooks/with_unpaired/mutant_calling/csv_files/20210603_plates_1-13_barcodes_WT_'+str(len(BCs_with_empty_vcf))+'.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Separate indel variants from single base substitutions\n",
    "Here, want to:\n",
    "1) calculate quantity of BCs containing indels and store list of these BCs\n",
    "2) Select entries only containing non-indels\n",
    "3) Remove any additional BCs that contained non-indels but also indel entries from step 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 403,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-403-f4a72b6e99ba>:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  substitutions['barcode_codon_index'] = substitutions['i7_barcode'].astype(str) + \"_\" +substitutions['codon'].astype(str)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "3029"
      ]
     },
     "execution_count": 403,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get list of BCs containing indels\n",
    "indelBCsL = dataDF.loc[True].i7_barcode.values.tolist()\n",
    "\n",
    "#also store indels in own dataframe\n",
    "indels=dataDF.loc[True]\n",
    "\n",
    "# get DF containing only non indel substitutions\n",
    "dataDF_non_indels=dataDF.loc[False] \n",
    "\n",
    "#select only entries with barcodes that are not in the list of BCs containing at least one indel\n",
    "substitutions=dataDF_non_indels.loc[~dataDF_non_indels.i7_barcode.isin(indelBCsL)] \n",
    "\n",
    "#change index for substitutions dataframe\n",
    "substitutions.reset_index(drop = False, inplace =True)\n",
    "substitutions['barcode_codon_index'] = substitutions['i7_barcode'].astype(str) + \"_\" +substitutions['codon'].astype(str)\n",
    "\n",
    "#get unique number of i7 barcodes\n",
    "substitutions['i7_barcode'].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Get amino acid change for single base substitution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 404,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>codon</th>\n",
       "      <th>refSeq</th>\n",
       "      <th>altSeq</th>\n",
       "      <th>filter</th>\n",
       "      <th>fw_WT</th>\n",
       "      <th>rv_WT</th>\n",
       "      <th>fw_alt</th>\n",
       "      <th>rv_alt</th>\n",
       "      <th>position</th>\n",
       "      <th>i7_barcode</th>\n",
       "      <th>barcode_codon_index</th>\n",
       "      <th>amino_acid_sub</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode_codon_index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">774_24</th>\n",
       "      <th>5048</th>\n",
       "      <td>24</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>225.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>403</td>\n",
       "      <td>257</td>\n",
       "      <td>142</td>\n",
       "      <td>774</td>\n",
       "      <td>774_24</td>\n",
       "      <td>L_24_V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5049</th>\n",
       "      <td>24</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>225.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>400</td>\n",
       "      <td>256</td>\n",
       "      <td>144</td>\n",
       "      <td>774</td>\n",
       "      <td>774_24</td>\n",
       "      <td>L_24_V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         codon refSeq altSeq  filter  fw_WT  rv_WT  fw_alt  \\\n",
       "barcode_codon_index                                                          \n",
       "774_24              5048    24      C      G   225.0      0      0     403   \n",
       "                    5049    24      G      T   225.0      0      0     400   \n",
       "\n",
       "                          rv_alt  position  i7_barcode barcode_codon_index  \\\n",
       "barcode_codon_index                                                          \n",
       "774_24              5048     257       142         774              774_24   \n",
       "                    5049     256       144         774              774_24   \n",
       "\n",
       "                         amino_acid_sub  \n",
       "barcode_codon_index                      \n",
       "774_24              5048         L_24_V  \n",
       "                    5049         L_24_V  "
      ]
     },
     "execution_count": 404,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#cast codon column as object (as opposed to int) if needed here\n",
    "substitutions=substitutions.astype({'codon': 'object'})\n",
    "substitutions.dtypes\n",
    "\n",
    "#ADD AMINO ACID SUB VALUE TO DATAFRAME WITH SINGLE BASE SUBSTITUTIONS ONLY\n",
    "df_with_subs=substitutions.groupby(['barcode_codon_index']).apply(get_amino_acid_change)\n",
    "df_with_subs.loc[df_with_subs['i7_barcode']==774]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. Add column containing ratio of variant:WT reads for each codon sub  \n",
    "EDIT: average the (F+R) sum of variant and WT reads across all nt subtitutions for a given codon sub \n",
    "-This will preserve the effects of having slightly different read counts for nt subs withing the same AA sub \n",
    "-Add columns containing sums of F and R reads for WT and var  \n",
    "-Add column with ratio of var:WT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 405,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>codon</th>\n",
       "      <th>refSeq</th>\n",
       "      <th>altSeq</th>\n",
       "      <th>filter</th>\n",
       "      <th>fw_WT</th>\n",
       "      <th>rv_WT</th>\n",
       "      <th>fw_alt</th>\n",
       "      <th>rv_alt</th>\n",
       "      <th>position</th>\n",
       "      <th>i7_barcode</th>\n",
       "      <th>barcode_codon_index</th>\n",
       "      <th>amino_acid_sub</th>\n",
       "      <th>WT_sum</th>\n",
       "      <th>var_sum</th>\n",
       "      <th>ratio_var</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode_codon_index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6135_384</th>\n",
       "      <th>1640</th>\n",
       "      <td>384</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>7.30814</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1223</td>\n",
       "      <td>6135</td>\n",
       "      <td>6135_384</td>\n",
       "      <td>Y_384_S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         codon refSeq altSeq   filter  fw_WT  rv_WT  fw_alt  \\\n",
       "barcode_codon_index                                                           \n",
       "6135_384            1640   384      A      C  7.30814      0      0       1   \n",
       "\n",
       "                          rv_alt  position  i7_barcode barcode_codon_index  \\\n",
       "barcode_codon_index                                                          \n",
       "6135_384            1640       0      1223        6135            6135_384   \n",
       "\n",
       "                         amino_acid_sub  WT_sum  var_sum  ratio_var  \n",
       "barcode_codon_index                                                  \n",
       "6135_384            1640        Y_384_S       0        1        inf  "
      ]
     },
     "execution_count": 405,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#copy DF from step 7\n",
    "df_with_subs_ratio=df_with_subs.copy()\n",
    "\n",
    "aveVarSumL=[]\n",
    "aveWTSumL=[]\n",
    "\n",
    "for index, val in df_with_subs_ratio.iterrows():\n",
    "    #average FWD and REV read counts (for variant and WT sequences) at variant position, then sum\n",
    "    var_sum_ave=int((sum((df_with_subs.loc[index[0]])['fw_alt'].tolist())/2) + (sum((df_with_subs.loc[index[0]])['rv_alt'].tolist())/2))\n",
    "    WT_sum_ave=int((sum((df_with_subs.loc[index[0]])['fw_WT'].tolist())/2) + (sum((df_with_subs.loc[index[0]])['rv_WT'].tolist())/2))\n",
    "    \n",
    "    #account for round down averages that produce variant sum averages that =0\n",
    "    if var_sum_ave == 0:\n",
    "        var_sum_ave= 1\n",
    "    else:\n",
    "        pass\n",
    "    \n",
    "    #add to lists\n",
    "    aveVarSumL.append(var_sum_ave)\n",
    "    aveWTSumL.append(WT_sum_ave)\n",
    "    \n",
    "\n",
    "#add averaged sum columns to DF\n",
    "df_with_subs_ratio['WT_sum']=aveWTSumL\n",
    "df_with_subs_ratio['var_sum']=aveVarSumL\n",
    "\n",
    "#add column containing ratio\n",
    "df_with_subs_ratio['ratio_var']=round((df_with_subs_ratio['var_sum']/df_with_subs_ratio['WT_sum']),2)\n",
    "\n",
    "\n",
    "df_with_subs_ratio.loc[df_with_subs_ratio['i7_barcode']==6135]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 9. Drop multiple positions changes listed for each barcode, just want codon change here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 406,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>codon</th>\n",
       "      <th>refSeq</th>\n",
       "      <th>altSeq</th>\n",
       "      <th>filter</th>\n",
       "      <th>fw_WT</th>\n",
       "      <th>rv_WT</th>\n",
       "      <th>fw_alt</th>\n",
       "      <th>rv_alt</th>\n",
       "      <th>position</th>\n",
       "      <th>i7_barcode</th>\n",
       "      <th>barcode_codon_index</th>\n",
       "      <th>amino_acid_sub</th>\n",
       "      <th>WT_sum</th>\n",
       "      <th>var_sum</th>\n",
       "      <th>ratio_var</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode_codon_index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>774_24</th>\n",
       "      <th>5048</th>\n",
       "      <td>24</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>225.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>403</td>\n",
       "      <td>257</td>\n",
       "      <td>142</td>\n",
       "      <td>774</td>\n",
       "      <td>774_24</td>\n",
       "      <td>L_24_V</td>\n",
       "      <td>0</td>\n",
       "      <td>658</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         codon refSeq altSeq  filter  fw_WT  rv_WT  fw_alt  \\\n",
       "barcode_codon_index                                                          \n",
       "774_24              5048    24      C      G   225.0      0      0     403   \n",
       "\n",
       "                          rv_alt  position  i7_barcode barcode_codon_index  \\\n",
       "barcode_codon_index                                                          \n",
       "774_24              5048     257       142         774              774_24   \n",
       "\n",
       "                         amino_acid_sub  WT_sum  var_sum  ratio_var  \n",
       "barcode_codon_index                                                  \n",
       "774_24              5048         L_24_V       0      658        inf  "
      ]
     },
     "execution_count": 406,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#for codon substitutions originating base changes at more than one position, drop\n",
    "#additional positions\n",
    "df_with_subs_culled_BC_C_index = df_with_subs_ratio.drop_duplicates(subset=['barcode_codon_index'])\n",
    "\n",
    "\n",
    "df_with_subs_culled_BC_C_index.loc[df_with_subs_culled_BC_C_index['i7_barcode']==774]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add plate and well information to each entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 387,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ##import csv containing i7/i5 barcodes, experimental plate number, and wells\n",
    "# ##plates 1-13\n",
    "# barcode_key1to13=pd.read_csv('/Users/mja/Appel_MutSequencing/Barcodes/20201212_MJA_1-13_plate_well_key.csv')\n",
    "# ##plates 14\n",
    "# # barcode_key14=pd.read_csv('/Users/mja/Appel_MutSequencing/Barcodes/20201212_MJA_14_plate_well_key.csv')\n",
    "\n",
    "# #select only plates 1-13 (since 14 contains duplicate barcodes and was sequenced separately)\n",
    "# plates1to13_barcodes=barcode_key1to13.loc[barcode_key1to13['plate'].isin(range(1,14))]\n",
    "# plates1to13_barcodes_i7_only=plates1to13_barcodes[['i7','plate','well']].copy()\n",
    "\n",
    "# #select barcodes for plate 14 only\n",
    "# barcodes14=barcode_key14.loc[barcode_key14['plate'] == 14]\n",
    "# plates14_barcodes_i7_only=barcodes14[['i7','plate','well']].copy()\n",
    "\n",
    "# ##merge plate1-13 barcode list with summary dataframe\n",
    "# df_with_subs_culled_BC_C_index = pd.merge(left=df_with_subs_culled_BC_C_index, right=plates1to13_barcodes_i7_only, left_on='i7_barcode', right_on='i7')\n",
    "\n",
    "# ##repeat for plate 14\n",
    "# # plate14_df_wells = pd.merge(left=plate14_df, right=plates14_barcodes_i7_only, left_on='i7', right_on='i7')\n",
    "\n",
    "# df_with_subs_culled_BC_C_index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 10. Add plate and well data to aggregated Plates 1-13 dataframes, export, then split into plates and export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import csv containing i7/i5 barcodes, experimental plate number, and wells\n",
    "barcode_key=pd.read_csv('/Users/mja/Appel_MutSequencing/Barcodes/20201212_MJA_1-13_plate_well_key.csv')\n",
    "\n",
    "#select only plates 1-6 (since 7 contains duplicate barcodes and was sequenced separately)\n",
    "plates=range(1,14)\n",
    "barcodes=barcode_key.loc[barcode_key['plate'].isin(plates)]\n",
    "barcodes_i7_only=barcodes[['i7','plate','well']].copy()\n",
    "\n",
    "#edit column name to match column name for barcodes used in dataframes within this notebook\n",
    "barcodes_i7_only.rename(columns={\"i7\": \"i7_barcode\"}, inplace=True)\n",
    "\n",
    "#list of dictionaries to add plate and well information to\n",
    "list_of_summary_DFs=[dataDF,indels,df_with_subs_culled_BC_C_index]\n",
    "\n",
    "#specify list of names to use as keys\n",
    "list_of_df_names=[\"mutant_calling_initial_dataframe_all_vars\",\n",
    "                  \"mutant_calling_indels\",\n",
    "                  \"mutant_calling_all_AA_subs\"]\n",
    "\n",
    "#create dictionary for substitutions in codon\n",
    "#zip function accepts list of keys and list of values to do this\n",
    "df_dict= dict(zip(list_of_df_names, list_of_summary_DFs))\n",
    "\n",
    "\n",
    "#make empty dict to add plate+well modified dataframes to\n",
    "df_dict_plate_well={}\n",
    "\n",
    "#make another empty dict to store dfs split by plate\n",
    "split_df_dict_plate_well={}\n",
    "\n",
    "\n",
    "#iterate through dictionary of dataframes\n",
    "for df_name, df in df_dict.items():\n",
    "    \n",
    "    #make new dataframe to store merged df\n",
    "    df_with_plate_well=pd.DataFrame()\n",
    "    \n",
    "    #merge on i7_barcode\n",
    "    df_with_plate_well= pd.merge(left=df, right=barcodes_i7_only, left_on='i7_barcode', right_on='i7_barcode')\n",
    "    #add df to dict\n",
    "    df_dict_plate_well[df_name]=df_with_plate_well\n",
    "    \n",
    "    #export plate aggregated DF\n",
    "    fullDF_P = Path('/Users/mja/Appel_MutSequencing/2021_mut_seq_workup_II/analyses_from_py_notebooks/with_unpaired/mutant_calling/csv_files/')\n",
    "    #export dataframe to csv using dynamic naming to specify plate\n",
    "    df_with_plate_well.to_csv(Path(fullDF_P, \"plates_1-13_\"+df_name+'.csv'))\n",
    "    \n",
    "    \n",
    "    #for each plate # in plate list (specified at top), create new dataframe containing rows with only that plate number\n",
    "    for i in plates:\n",
    "        split_df=df_with_plate_well.loc[df_with_plate_well['plate'] == i]\n",
    "        \n",
    "        #create key name to add this new dataframe to a dictionary to access later if needed\n",
    "        df_name_with_plate=\"plate_\"+str(i)+\"_\"+df_name\n",
    "        \n",
    "        #add to dictionary\n",
    "        split_df_dict_plate_well[df_name_with_plate]=split_df\n",
    "        \n",
    "        from pathlib import Path\n",
    "        \n",
    "        #specify output directory\n",
    "        p = Path('/Users/mja/Appel_MutSequencing/2021_mut_seq_workup_II/analyses_from_py_notebooks/with_unpaired/mutant_calling/csv_files/plate-well/')\n",
    "        #export dataframe to csv using dynamic naming to specify plate\n",
    "        split_df.to_csv(Path(p, df_name_with_plate+'.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
